### PR TITLE
Transform tool dynamics.

### PR DESCRIPTION
**Pull request by bbastings**

### CODE REVIEW

File: editor/frontend/src/TransformElementsTool.ts [hunk 1]
```diff
@@ -3,12 +3,160 @@
...
+
+  private getRequestId(id: Id64String): string { return `${this.prefix}-${id}`; }
+  private getToleranceLog10(): number { return Math.log10(this.chordTolerance); }
+
+  private async createRequest(id: Id64String): Promise<TransformGraphicsData | undefined> {
+    const elementProps = await this.iModel.elements.getProps(id);
+    if (0 === elementProps.length)
+      return;
+
+    const placementProps = (elementProps[0] as any).placement;
```
Comment (kabentley) [inline editor/frontend/src/TransformElementsTool.ts:45]: `as any`, yuk. `GeometricElement3dProps | GeometricElement2dProps`?

File: editor/frontend/src/TransformElementsTool.ts [hunk 2]
```diff
@@ -3,12 +3,160 @@
...
+    if (undefined === placementProps)
+      return;
+
+    const hasAngle = (arg: any): arg is Placement2dProps => arg.angle !== undefined;
```
Comment (kabentley) [inline editor/frontend/src/TransformElementsTool.ts:49]: `const is2d = placementProps.angle !== undefined`

File: editor/frontend/src/TransformElementsTool.ts [hunk 3]
```diff
@@ -3,12 +3,160 @@
...
+      toleranceLog10: this.getToleranceLog10(),
+    };
+
+    this.pending.set(id, requestProps.id); // keep track of requests so they can be cancelled...
+
+    const graphicData = await IModelApp.tileAdmin.requestElementGraphics(this.iModel, requestProps);
+    if (undefined === graphicData)
+      return;
+
+    const graphic = await readElementGraphics(graphicData, this.iModel, elementProps[0].model, placement instanceof Placement2d ? false : true);
```
Comment (kabentley) [inline editor/frontend/src/TransformElementsTool.ts:67]: `is2d` rather than instanceof

File: editor/frontend/src/TransformElementsTool.ts [hunk 4]
```diff
@@ -138,6 +267,12 @@ export abstract class TransformElementsTool extends ElementSetTool {
       return; // TODO: Update agenda to hold copies, replace current selection set with copies, etc...
     return super.onProcessComplete();
   }
+
+  public onCleanup(): void {
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.clearAgendaGraphics();
```
Comment (kabentley) [inline editor/frontend/src/TransformElementsTool.ts:273]: @pmconne is it ok to not wait for this?

File: editor/frontend/src/TransformElementsTool.ts [hunk 5]
```diff
@@ -323,20 +458,21 @@ export class RotateElementsTool extends TransformElementsTool {
...
       return super.transformAgendaDynamics(transform, context);
 
-    if (undefined === this._elementAlignedBoxes || undefined === this._elementOrigins || this._elementAlignedBoxes.length !== this._elementOrigins.length)
+    if (undefined === this._graphicsProvider)
       return;
 
-    const builder = context.target.createGraphicBuilder(GraphicType.WorldDecoration, context.viewport);
-    builder.setSymbology(context.viewport.getContrastToBackgroundColor(), ColorDef.black, 1, LinePixels.HiddenLine);
+    for (const data of this._graphicsProvider.data) {
+      const rotatePoint = Point3d.create();
```
Comment (kabentley) [inline editor/frontend/src/TransformElementsTool.ts:465]: maybe take this out of the loop? Same thing for rotateTrans.

**Review by kabentley — CHANGES_REQUESTED** [review]

File: editor/frontend/src/TransformElementsTool.ts [hunk 6]
```diff
@@ -3,12 +3,160 @@
...
+
+  constructor(iModel: IModelConnection, prefix: string) {
+    this.iModel = iModel;
+    this.prefix = prefix;
+    this.data = new Array<TransformGraphicsData>();
+    this.pending = new Map<Id64String, string>();
+  }
+
+  private getRequestId(id: Id64String): string { return `${this.prefix}-${id}`; }
+  private getToleranceLog10(): number { return Math.log10(this.chordTolerance); }
```
Comment (pmconne) [inline editor/frontend/src/TransformElementsTool.ts:38]: Native code will currently reject your tolerance if it is not an integer.

File: editor/frontend/src/TransformElementsTool.ts [hunk 7]
```diff
@@ -3,12 +3,160 @@
...
+    const branch = new GraphicBranch(false);
+    branch.add(graphic);
+
+    const branchGraphic = context.createBranch(branch, transform);
+    context.addGraphic(branchGraphic);
+  }
+
+  public addGraphics(transform: Transform, context: DynamicsContext): void {
+    for (const data of this.data)
+      this.addSingleGraphic(data.graphic, transform, context);
```
Comment (pmconne) [inline editor/frontend/src/TransformElementsTool.ts:157]: Considered putting all the graphics into a single branch?

File: editor/frontend/src/TransformElementsTool.ts (hunk 2 above)
Comment (bbastings) [reply editor/frontend/src/TransformElementsTool.ts:49]: Property 'angle' does not exist on type 'Placement3dProps | Placement2dProps'.
  Property 'angle' does not exist on type 'Placement3dProps'.ts(2339)

zoomToPlacementProps has placement as any...

Comment (kabentley) [reply editor/frontend/src/TransformElementsTool.ts:49]: `const is2d = (placementProps as any).angle !== undefined`

File: editor/frontend/src/TransformElementsTool.ts (hunk 1 above)
Comment (bbastings) [reply editor/frontend/src/TransformElementsTool.ts]: done.

File: editor/frontend/src/TransformElementsTool.ts (hunk 5 above)
Comment (bbastings) [reply editor/frontend/src/TransformElementsTool.ts]: done.

File: editor/frontend/src/TransformElementsTool.ts (hunk 6 above)
Comment (bbastings) [reply editor/frontend/src/TransformElementsTool.ts]: Added floor.

File: editor/frontend/src/TransformElementsTool.ts (hunk 7 above)
Comment (bbastings) [reply editor/frontend/src/TransformElementsTool.ts]: done.

**Review by pmconne — APPROVED** [review]

File: core/common/src/geometry/Placement.ts [hunk 8]
```diff
@@ -40,6 +40,8 @@ export type ElementAlignedBox2d = Range2d;
  */
 export type LocalAlignedBox3d = Range3d;
 
+export type Placement = Placement2d | Placement3d;
```
Comment (pmconne) [inline core/common/src/geometry/Placement.ts:43]: Add to `GeometricElementProps` a `placement: Placement` property?

Comment (kabentley) [reply core/common/src/geometry/Placement.ts:46]: yeah, that's probably better

Comment (kabentley) [reply core/common/src/geometry/Placement.ts:46]: done

**Review by kabentley — APPROVED** [review]

**Pull request merged**